passes/opt/opt_demorgan.cc

```diff
@@ -0,0 +1,202 @@
+			{
+				srcinv = x.cell;
+				break;
+			}
+		}
+
+		//We are NOT inverted! Add an inverter
+		if(!srcinv)
+		{
+			auto inverted_b = m->addWire(NEW_ID);
+			m->addNot(NEW_ID, RTLIL::SigSpec(b), RTLIL::SigSpec(inverted_b));
+			insig[i] = inverted_b;
+		}
+
+		//We ARE inverted - bypass it
+		//Don't automatically delete the inverter since other stuff might still use it
+		else
+			insig[i] = srcinv->getPort("\\A");
+	}
+
+	//Cosmetic fixup: If our input is just a scrambled version of one bus, rearrange it
+	//Reductions are all commutative, so there's no point in having them in a weird order
+	bool same_signal = true;
+	RTLIL::Wire* srcwire = insig[0].wire;
+	std::map<int, int> seen_bits;
+	for(int i=0; i<insig.size(); i++)
+		seen_bits[i] = 0;
+	for(int i=0; i<insig.size(); i++)
+	{
+		seen_bits[insig[i].offset] ++;
+		if(insig[i].wire != srcwire)
+		{
+			same_signal = false;
+			break;
+		}
+	}
+	if(same_signal)
+	{
+		//Make sure we've seen every bit exactly once
+		bool every_bit_once = true;
+		for(int i=0; i<insig.size(); i++)
+		{
+			if(seen_bits[i] != 1)
+			{
+				every_bit_once = false;
+				break;
+			}
+		}
+
+		//All good? Just use the whole wire as-is without any reordering
+		//We do have to swap MSB to LSB b/c that's the way the reduction cells seem to work?
+		//Unclear on why this isn't sorting properly
+		//TODO: can we do SigChunks instead of single bits if we have subsets of a bus?
+		if(every_bit_once && (insig.size() == srcwire->width) )
+		{
+			log("Rearranging bits\n");
+			RTLIL::SigSpec newsig;
+			for(int i=0; i<insig.size(); i++)
+				newsig.append(RTLIL::SigBit(srcwire, insig.size() - i - 1));
+			insig = newsig;
+			insig.sort();
+		}
+	}
+
+	//Push the new input signal back to the reduction (after bypassing/adding inverters)
+	cell->setPort("\\A", insig);
+
+	//Change the cell type
+	if(cell->type == "$reduce_and")
+		cell->type = "$reduce_or";
+	else if(cell->type == "$reduce_or")
+		cell->type = "$reduce_and";
+	//don't change XOR
+
+	//Add an inverter to the output
+	auto inverted_output = cell->getPort("\\Y");
+	auto uninverted_output = m->addWire(NEW_ID);
+	m->addNot(NEW_ID, RTLIL::SigSpec(uninverted_output), inverted_output);
+	cell->setPort("\\Y", uninverted_output);
+}
+
+struct OptDemorganPass : public Pass {
+	OptDemorganPass() : Pass("opt_demorgan", "Optimize reductions with DeMorgan equivalents") { }
+	virtual void help()
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    opt_demorgan [selection]\n");
+		log("\n");
+		log("This pass pushes inverters through $reduce_* cells if this will reduce the\n");
+		log("overall gate count of the circuit\n");
+		log("\n");
+	}
+	virtual void execute(std::vector<std::string> /*args*/, RTLIL::Design *design)
+	{
+		log_header(design, "Executing OPT_DEMORGAN pass (push inverters through $reduce_* cells).\n");
+
+		//int argidx = 0;
+		//extra_args(args, argidx, design);
```

Comment:
Without `extra_args()` the `[selection]` in the help message is a lie. :)
Please call `extra_args()`.